src/nlua_shiplog.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * See Licensing and Copyright notice in naev.h
+ */
+
+/**
+ * @file nlua_shiplog.c
+ *
+ * @brief Handles the shiplog Lua bindings.
+ */
+
+
+#include "nlua_shiplog.h"
+
+#include "naev.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include "nstring.h"
+#include <math.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+
+#include "nlua.h"
+#include "nlua_hook.h"
+#include "nlua_player.h"
+#include "nlua_tk.h"
+#include "nlua_faction.h"
+#include "nlua_system.h"
+#include "nlua_tex.h"
+#include "nlua_camera.h"
+#include "nlua_music.h"
+#include "nlua_bkg.h"
+#include "nlua_misn.h"
+#include "player.h"
+#include "mission.h"
+#include "log.h"
+#include "rng.h"
+#include "toolkit.h"
+#include "land.h"
+#include "nxml.h"
+#include "nluadef.h"
+#include "music.h"
+#include "gui_osd.h"
+#include "npc.h"
+#include "array.h"
+#include "ndata.h"
+#include "shiplog.h"
+
+int shiplog_loadShiplog( nlua_env env );
+static int shiplog_createLog( lua_State *L );
+static int shiplog_appendLog( lua_State *L );
+static const luaL_Reg shiplog_methods[] = {
+   { "createLog", shiplog_createLog },
+   { "appendLog", shiplog_appendLog },
+   {0,0}
+}; /**< Shiplog Lua methods. */
+
+
+/**
+ * @brief Registers the lua libraries.
+ *
+ *    @param L Lua state.
+ *    @return 0 on success.
+ */
+int shiplog_loadLibs( nlua_env env )
+{
+   nlua_loadStandard(env);
+   nlua_loadShiplog(env);
+   return 0;
+}
+/*
+ * individual library loading
+ */
+/**
+ * @brief Loads the mission Lua library.
+ *    @param L Lua state.
+ */
+int nlua_loadShiplog( nlua_env env )
+{
+   nlua_register(env, "shiplog", shiplog_methods, 0);
+   return 0;
+}
+
+
+/**
+ * @brief Creates a shiplog for this mission.
+ *
+ *    @luatparam string name Name for this log.
+ *    @luatparam string type Type of log (e.g travel, trade, etc, can be anything.)
+ *    @luatparam number overwrite Whether to overwrite existing mission with this logname and logtype.  Warning, removes previous entries of this logname and type if 1, or logs of this type if 2.
+ *    @luatreturn number The logid of the mission.
+ * @luafunc createLog( logname, logtype, overwrite )
+ * @usage misn.createLog("My mission title","Mission type",0)
+ * @usage misn.createLog("Any title","Anything can be a type",1) with 1 to replace existing missions of this title and type.
+ * @usage misn.createLog("Any title","Anything can be a type",2) with 2 to replace existing missions of type.
+ */
+static int shiplog_createLog( lua_State *L )
+{
+   const char *logname;
+   const char *logtype;
+   int overwrite,logid;
+   /* Parameters. */
+   logname    = luaL_checkstring(L,1);
+   logtype    = luaL_checkstring(L,2);
+   overwrite = luaL_checkint(L,3);
+
+   /* Create a new shiplog */
+   logid = shiplog_create( logname, logtype, overwrite );
```

Comment:
It looks like you're using the display names of the log name and log type for differentiation? This is a bad idea, it would be a repeat of the problem we have with news articles. It would be better to have separate strings - ID strings, an additional two arguments - that identify categorization on a programmatic level. The pretty version, which would be translated, should only be used for display.

🕵️

Reply:
No - the logid is used for differentiation, once a log has been created.
The log type and logname are displayed to the user in the info widget, but thats all - and these should be translated.
No other differentiation is required.

Reply:
The logid is independent of the logname and logtype...

Reply:
Then how do multiple, separate Lua scripts use the same "log"? I've told you this already, there needs to be a way for the Lua scripter to specify *where* the log goes. A numeric, auto-generated "log ID" doesn't cut it because multiple scripts (whether that's multiple missions or any other situations) cannot easily share these. You need *string* identifiers.

🕵️

Reply:
ok - now I see - I wasn't aware of the need for this before.  I have a solution - will push it this evening, but basically, a new createNamedLog function which will create a logID if needed, or return it if it already exists, and that will solve it.

Reply:
That would work, and I would suggest just making that the only logging function, so that everything uses that. Having one without the feature doesn't really add anything, it would just create confusion.

Reply:
The reason to keep it would be for missions that can be done more than once - and we don't know whether these will implement logs in the future or not - so for now, I think best to keep it.  However, I'll get rid of it if you disagree.

Reply:
Why can't the missions just make a string up that they'll use? It could be unique and that would be fine. If you have one function it unifies everything and makes it crystal-clear to scripters where commonalities lie as far as categorization.

Reply:
Right, I'll do that... will push later.

Reply:
ok - now pushed - updated the createLog function to allow separate lua scripts get the same logset.

Reply:
Awesome, thanks!

🕷️